main_test.go

```diff
@@ -512,10 +515,71 @@ var testCases = []TestCase{
 	},
 }
 
+type FakeS3Client struct {
+	Get mockGetObjectAPI
+	Put mockPutObjectAPI
+}
+
+type mockGetObjectAPI func(ctx context.Context, params *s3.GetObjectInput) (*s3.GetObjectOutput, error)
+type mockPutObjectAPI func(ctx context.Context, params *s3.PutObjectInput) (*s3.PutObjectOutput, error)
+
+func (fc *FakeS3Client) GetObject(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.Options)) (*s3.GetObjectOutput, error) {
+	return fc.Get(ctx, params)
+}
+
+func (fc *FakeS3Client) PutObject(ctx context.Context, params *s3.PutObjectInput, optFns ...func(*s3.Options)) (*s3.PutObjectOutput, error) {
+	return fc.Put(ctx, params)
+}
+
+func (fc *FakeS3Client) New(bucket, key string) *FakeS3Client {
```

Comment:
sorry, I forgot to write this in a previous review:

I think it's a somewhat odd API to have to make a `FakeS3Client`, and then call `New` on it to get a *new* `FakeS3Client`. And I'm not sure there's any need for the additional layer of indirection to have a type with function fields and a closure containing data instead of just methods on a type that refer to fields containing data.

What about

```
type FakeS3Client struct {
	Bucket, Key string
}

func (fc *FakeS3Client) GetObject(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.Options)) (*s3.GetObjectOutput, error) {
	// implementation from "New" but using fc.Bucket and fc.Key instead of bucket and key
}
```

Reply:
Yes. That is exactly what I did (before reading your suggestion). Again, much simpler.